doc/man3/EVP_PKEY_fromdata.pod

```diff
@@ -120,52 +120,67 @@ TODO Write a set of cookbook documents and link to them.
 =head2 Creating an ECC keypair using raw key data
 
  #include <openssl/evp.h>
+ #include <openssl/param_build.h>
 
  /*
-  * These arrays represent large numbers, big endian organization.
-  * In a real application, these would probably be bignums that get
-  * converted to the native integer organization with BN_bn2nativepad().
-  * We're not doing that here, since this is not an example of BIGNUM
-  * functionality, but an example of EVP_PKEY_fromdata().
+  * Fix data to represent the private and public key.
   */
- #ifndef B_ENDIAN
- # error "We haven't prepared little endian arrays"
- #endif
- const unsigned char priv[] = {
+ const unsigned char priv_data[] = {
      0xb9, 0x2f, 0x3c, 0xe6, 0x2f, 0xfb, 0x45, 0x68,
      0x39, 0x96, 0xf0, 0x2a, 0xaf, 0x6c, 0xda, 0xf2,
      0x89, 0x8a, 0x27, 0xbf, 0x39, 0x9b, 0x7e, 0x54,
      0x21, 0xc2, 0xa1, 0xe5, 0x36, 0x12, 0x48, 0x5d
  };
- const unsigned char pub[] = {
-     0x04, 0xcf, 0x20, 0xfb, 0x9a, 0x1d, 0x11, 0x6c,
-     0x5e, 0x9f, 0xec, 0x38, 0x87, 0x6c, 0x1d, 0x2f,
-     0x58, 0x47, 0xab, 0xa3, 0x9b, 0x79, 0x23, 0xe6,
-     0xeb, 0x94, 0x6f, 0x97, 0xdb, 0xa3, 0x7d, 0xbd,
-     0xe5, 0x26, 0xca, 0x07, 0x17, 0x8d, 0x26, 0x75,
-     0xff, 0xcb, 0x8e, 0xb6, 0x84, 0xd0, 0x24, 0x02,
-     0x25, 0x8f, 0xb9, 0x33, 0x6e, 0xcf, 0x12, 0x16,
-     0x2f, 0x5c, 0xcd, 0x86, 0x71, 0xa8, 0xbf, 0x1a,
-     0x47
- };
- const OSSL_PARAM params[] = {
-     OSSL_PARAM_utf8_string("curve-name", "prime256v1"),
-     OSSL_PARAM_BN("priv", priv, sizeof(priv)),
-     OSSL_PARAM_BN("pub", pub, sizeof(pub)),
-     OSSL_PARAM_END
+ /* UNCOMPRESSED FORMAT */
+ const unsigned char pub_data[] = {
+     POINT_CONVERSION_UNCOMPRESSED,
+     0xcf, 0x20, 0xfb, 0x9a, 0x1d, 0x11, 0x6c, 0x5e,
+     0x9f, 0xec, 0x38, 0x87, 0x6c, 0x1d, 0x2f, 0x58,
+     0x47, 0xab, 0xa3, 0x9b, 0x79, 0x23, 0xe6, 0xeb,
+     0x94, 0x6f, 0x97, 0xdb, 0xa3, 0x7d, 0xbd, 0xe5,
+     0x26, 0xca, 0x07, 0x17, 0x8d, 0x26, 0x75, 0xff,
+     0xcb, 0x8e, 0xb6, 0x84, 0xd0, 0x24, 0x02, 0x25,
+     0x8f, 0xb9, 0x33, 0x6e, 0xcf, 0x12, 0x16, 0x2f,
+     0x5c, 0xcd, 0x86, 0x71, 0xa8, 0xbf, 0x1a, 0x47
  };
 
  int main()
  {
-     EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL);
+     EVP_PKEY_CTX *ctx;
      EVP_PKEY *pkey = NULL;
-
+     BIGNUM *priv;
+     OSSL_PARAM_BLD *param_bld;
+     OSSL_PARAM *params = NULL;
+     int exitcode = 0;
+
+     priv = BN_bin2bn(priv_data, sizeof(priv_data), NULL);
+
+     param_bld = OSSL_PARAM_BLD_new();
+     if (priv != NULL && param_bld != NULL
+         && OSSL_PARAM_BLD_push_utf8_string(param_bld, "group",
```

Comment:
Once we stop using the `OSSL_PARAM`s directly and start using the `OSSL_PARAM_BLD` framework does it still make sense to use `"group"`, `"priv"`, `"pub"` directly instead of the appropriate name definitions?

Reply:
Is there a sane way of passing these 3 parameters (which are just a string and 2 octet buffers) using `OSSL_PARAM` directly?

It seems from the initial "simplicity" and portability of the `OSSL_PARAM` interface we have diverged quite a lot!

Reply:
If you read [OSSL_PARAM(3)](https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_PARAM.pod), you will find a few raw examples.  They should work in C99 and newer at least.

Otherwise, it's perfectly possible, although lengthier, do be even more raw (the same as you would with PKCS#11 templates):

``` C
const OSSL_PARAM params[4];

params[0].key = "curve-name";
params[0].data_type = OSSL_PARAM_UTF8_STRING;
params[0].data = (void *) "prime256v1";
params[0].data_size = sizeof("prime256v1");
params[0].return_size = 0;
params[1].key = "priv";
params[1].data_type = OSSL_PARAM_UNSIGNED_INTEGER;
params[1].data = priv;
params[1].data_size = sizeof(priv);
params[1].return_size = 0;
params[2].key = "pub";
params[2].data_type = OSSL_PARAM_OCTET_STRING;
params[2].data = pub;
params[2].data_size = sizeof(pub);
params[2].return_size = 0;
params[3].key = NULL;
params[3].data_type = 0;
params[3].data = NULL;
params[3].data_size = 0;
params[3].return_size = 0;
```

`OSSL_PARAM` *is*, after all, an open structure.  This is by design.

Reply:
> Once we stop using the `OSSL_PARAM`s directly and start using the `OSSL_PARAM_BLD` framework does it still make sense to use `"group"`, `"priv"`, `"pub"` directly instead of the appropriate name definitions?
ren
I'm pushing for this to avoid adding a sense that OSSL_blah_PARAM_something macros are mandatory.  This is to avoid registration hell like we've had around `crypto/objects/objects.txt` and NIDs.

Reply:
@levitte thanks for the clarification!

Reply:
The downside to this is that when someone decides to change a name like "named_curve" to "group" then all of the pod file examples get out of sync. It has happened in quite a few places.
I would suggest this is the entire reason for the defines, and that you are shooting yourself in the foot using the direct names.